File: golang/Arrays/arrays.go
package arrays

import "math"

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func Reverse(nums []int) []int {
	var i int = 0
	var j int = len(nums) - 1
	for i < j {
		nums[i], nums[j] = nums[j], nums[i]
		i++
		j--
	}
	return nums
}

func LinearSearch(nums []int, key int) int {
	for i, v := range nums {
		if v == key {
			return i
		}
	}
	return -1
}

func SearchRange(nums []int, target int) [2]int {
	answer := [2]int{-1, -1}
	for i := 0; i < len(nums); i++ {
		if nums[i] == target {
			answer[0] = i
			break
		}
	}
	for i := len(nums) - 1; i > -1; i-- {
		if nums[i] == target {
			answer[1] = i
			break
		}
	}
	return answer
}

func CreateFrequencyTable(nums []int) map[int]int {
	frequencyTable := make(map[int]int)
	for _, v := range nums {
		if _, exist := frequencyTable[v]; exist {
			frequencyTable[v]++
		} else {
			frequencyTable[v] = 1
		}
	}
	return frequencyTable
}

func FindHighestFrequency(nums []int) int {
	frequencyTable := make(map[int]int)
	var maxFrequency int = 0
	for _, v := range nums {
		if _, exist := frequencyTable[v]; exist {
			frequencyTable[v]++
			maxFrequency = maxInt(frequencyTable[v], maxFrequency)
		} else {
			frequencyTable[v] = 1
		}
	}
	return maxFrequency
}

func FindLowestFrequency(nums []int) int {
	frequencyTable := make(map[int]int)
	var minFrequency int = math.MaxInt
	for _, v := range nums {
		if _, exist := frequencyTable[v]; exist {
			frequencyTable[v]++
		} else {
			frequencyTable[v] = 1
		}
	}
	for _, v := range frequencyTable {
		minFrequency = minInt(minFrequency, v)
	}
	return minFrequency
}

func FindLargest(nums []int) int {
	var largest int = nums[0]
	for _, v := range nums {
		if v > largest {
			largest = v
		}
	}
	return largest
}

func FindSmallest(nums []int) int {
	smallest := nums[0]
	for _, v := range nums {
		if v < smallest {
			smallest = v
		}
	}
	return smallest
}

func FindSecondLargest(nums []int) int {
	var largest int = nums[0]
	var secondLargest int = nums[0]
	for i := range nums {
		if nums[i] > largest {
			secondLargest = largest
			largest = nums[i]
		} else if nums[i] > secondLargest && nums[i] < largest {
			secondLargest = nums[i]
		}
	}
	return secondLargest
}

func FindSecondSmallest(nums []int) int {
	smallest := nums[0]
	secondSmallest := nums[0]
	for i := range nums {
		if nums[i] < smallest {
			secondSmallest = smallest
			smallest = nums[i]
		} else if nums[i] < secondSmallest && nums[i] > smallest {
			secondSmallest = nums[i]
		}
	}
	return secondSmallest
}

func Leaders(nums []int) []int {
	answer := []int{}
	max := nums[len(nums)-1]
	for i := len(nums) - 2; i >= 0; i-- {
		if nums[i] >= max {
			answer = append(answer, nums[i])
			max = nums[i]
		}
	}
	return Reverse(answer)
}

func FindSingleI(nums []int) int {
	frequencyTable := make(map[int]int)
	for _, v := range nums {
		if _, exist := frequencyTable[v]; exist {
			frequencyTable[v]++
		} else {
			frequencyTable[v] = 1
		}
	}
	for key, val := range frequencyTable {
		if val == 1 {
			return key
		}
	}
	return -1
}

func FindSingleII(nums []int) int {
	var xorSum int = 0
	for _, val := range nums {
		xorSum = xorSum ^ val
	}
	return xorSum
}

func CheckSorted(nums []int) bool {
	if len(nums) == 1 || len(nums) == 0 {
		return true
	}
	for i := 1; i <= len(nums)-1; i++ {
		if nums[i] < nums[i-1] {
			return false
		}
	}
	return true
}

func CheckRotatedSorted(nums []int) bool {
	var count int = 0
	for i := 1; i < len(nums); i++ {
		if nums[i] < nums[i-1] {
			count++
		}
	}
	if count == 0 {
		return true
	}
	if count == 1 && nums[0] >= nums[len(nums)-1] {
		return true
	}
	return false
}

func FindEvenDigits(nums []int) int {
	var digitCount int = 0
	var answer int = 0
	for i := range nums {
		var digit int = nums[i]
		digitCount = 0
		for digit > 0 {
			digit = digit / 10
			digitCount += 1
		}
		if digitCount%2 == 0 {
			answer += 1
		}
	}
	return answer
}

func MaxConsecutiveOnes(nums []int) int {
	var localCount int = 0
	var globalCount int = 0
	for i := range nums {
		if nums[i] == 1 {
			localCount += 1
		} else {
			localCount = 0
		}
		globalCount = maxInt(localCount, globalCount)
	}
	return globalCount
}

func FindMissing(nums []int, n int) int {
	var sum int = 0
	for i := range nums {
		sum += nums[i]
	}
	return n*(n+1)/2 - sum
}

func TwoSumI(nums []int, target int) [2]int {
	for i := range nums {
		for j := range nums {
			if nums[j]+nums[i] == target {
				return [2]int{i, j}
			}
		}
	}
	return [2]int{-1, -1}
}

func TwoSumII(nums []int, target int) [2]int {
	search := make(map[int]int)
	for i, v := range nums {
		if idx, exist := search[target-v]; exist {
			return [2]int{idx, i}
		} else {
			search[v] = i
		}
	}
	return [2]int{-1, -1}
}

func RemoveDuplicates(nums []int) int {
	var i int = 0
	for j := 1; j < len(nums); j++ {
		if nums[i] != nums[j] {
			i++
			nums[i] = nums[j]
		}
	}
	return i + 1
}

func RemoveZeros(nums []int) []int {
	var i int = 0
	for j := 0; j < len(nums); j++ {
		if nums[j] != 0 {
			nums[i], nums[j] = nums[j], nums[i]
			i++
		}
	}
	return nums
}

func ContainsDuplicate(nums []int) bool {
	search := make(map[int]bool)
	for _, v := range nums {
		if _, exist := search[v]; exist {
			return true
		} else {
			search[v] = true
		}
	}
	return false
}

func LeftRotateK(nums []int, k int) []int {
	k = k % len(nums)
	if k != 0 {
		temp := make([]int, len(nums))
		temp = append(nums[k:], nums[:k]...)
		copy(nums, temp)
	}
	return nums
}

func RightRotateK(nums []int, k int) []int {
	k = k % len(nums)
	if k != 0 {
		temp := make([]int, len(nums))
		temp = append(nums[len(nums)-k:], nums[:len(nums)-k]...)
		copy(nums, temp)
	}
	return nums
}

func LowerBound(nums []int, target int) int {
	var start int = 0
	var end int = len(nums) - 1
	var answer int = len(nums)
	for start <= end {
		var mid int = start + (end-start)/2
		if nums[mid] >= target {
			answer = mid
			end = mid - 1
		} else {
			start = mid + 1
		}
	}
	return answer
}

func SortColors(nums []int) {
	var left int = 0
	var mid int = 0
	var right int = len(nums) - 1
	for mid <= right {
		if nums[mid] == 0 {
			nums[mid], nums[left] = nums[left], nums[mid]
			left++
			mid++
		} else if nums[mid] == 2 {
			nums[right], nums[mid] = nums[mid], nums[right]
			right--
		} else {
			mid++
		}
	}
}

func MajorityElementI(nums []int) int {
	var count int = 0
	var element int
	for _, v := range nums {
		if count == 0 {
			element = v
			count++
		} else if v == element {
			count++
		} else {
			count--
		}
	}
	return element
}

func MajorityElementII(nums []int) []int {
	var c1 int = 0
	var c2 int = 0
	var ele1 int = 0
	var ele2 int = 0
	answer := []int{}
	for i := range nums {
		if nums[i] == ele1 {
			c1++
		} else if nums[i] == ele2 {
			c2++
		} else if c1 == 0 {
			ele1 = nums[i]
			c1++
		} else if c2 == 0 {
			ele2 = nums[i]
			c2++
		} else {
			c1--
			c2--
		}
	}
	c1 = 0
	c2 = 0
	for i := range nums {
		if nums[i] == ele1 {
			c1++
		} else if nums[i] == ele2 {
			c2++
		}
	}
	if c1 > len(nums)/3 {
		answer = append(answer, ele1)
	}
	if c2 > len(nums)/3 {
		answer = append(answer, ele2)
	}
	return answer
}

func MaxSubArraySumI(nums []int) int {
	var sum int = 0
	var max int = math.MinInt
	for _, v := range nums {
		sum = maxInt(sum+v, v)
		max = maxInt(sum, max)
	}
	return max
}

func MaxSubArraySumII(nums []int) []int {
	var sum int = 0
	var max int = math.MinInt
	var start int = 0
	var ansEnd, ansStart int = -1, -1
	for i := range nums {
		if sum == 0 {
			start = i
		}
		sum += nums[i]
		if sum > max {
			max = sum
			ansStart = start
			ansEnd = i
		}
		if sum <= 0 {
			sum = 0
		}
	}
	return nums[ansStart : ansEnd+1]
}

func MaxProfit(prices []int) int {
	var min int = math.MaxInt
	var diff int = -1
	for _, v := range prices {
		min = minInt(min, v)
		diff = maxInt(diff, v-min)
	}
	return diff
}

func RearrangeBySign(nums []int) []int {
	var positive int = 0
	var negative int = 1
	new := make([]int, len(nums))
	for i := range nums {
		if nums[i] >= 0 {
			new[positive] = nums[i]
			positive += 2
		} else {
			new[negative] = nums[i]
			negative += 2
		}
	}
	return new
}

func Merge(nums1, nums2 []int, m, n int) []int {
	var i int = m - 1
	var j int = n - 1
	var k int = m + n - 1
	for j >= 0 && i >= 0 {
		if nums1[i] > nums2[j] {
			nums1[k] = nums1[i]
			i--
		} else {
			nums1[k] = nums2[j]
			j--
		}
		k--
	}
	for j >= 0 {
		nums1[k] = nums2[j]
		j--
		k--
	}
	return nums1
}

func SubArraysWithSumI(nums []int, k int) int {
	var answer int = 0
	for i := range nums {
		var sum int = 0
		for j := i; j < len(nums); j++ {
			sum += nums[j]
			if sum == k {
				answer = maxInt(answer, j-i+1)
			}
		}
	}
	return answer
}

func SubArraysWithSumII(nums []int, k int) int {
	prefixSum := make(map[int]int)
	var sum int = 0
	var answer int = 0
	prefixSum[0] = 1
	for i := range nums {
		sum += nums[i]
		var remaining int = sum - k
		if _, ok := prefixSum[remaining]; ok {
			answer += prefixSum[remaining]
		}
		prefixSum[sum]++
	}
	return answer
}

func LongestSubArrayWithSumPositives(nums []int, k int) int {
	var i int = 0
	var answer int = 0
	var sum int = 0
	for j := range nums {
		sum += nums[j]
		for sum > k && i <= j {
			sum -= nums[i]
			i++
		}
		if sum == k {
			answer = maxInt(answer, j-i+1)
		}
	}
	return answer
}

func RotateMatrix(nums [][]int) [][]int {
	for i := range nums {
		for j := i; j < len(nums[0]); j++ {
			nums[i][j], nums[j][i] = nums[j][i], nums[i][j]
		}
	}
	for i := 0; i < len(nums); i++ {
		j, k := 0, len(nums[0])-1
		for j <= k {
			nums[i][j], nums[i][k] = nums[i][k], nums[i][j]
			j++
			k--
		}
	}
	return nums
}

func SetZeroes(matrix [][]int) [][]int {
	rows := make(map[int]bool)
	columns := make(map[int]bool)
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			if matrix[i][j] == 0 {
				rows[i] = true
				columns[j] = true
			}
		}
	}
	for i := range matrix {
		for j := range matrix[0] {
			if rows[i] || columns[j] {
				matrix[i][j] = 0
			}
		}
	}
	return matrix
}

func SpiralMatix(matrix [][]int) []int {
	var rows int = len(matrix)
	var columns int = len(matrix[0])
	var left int = 0
	var right int = columns - 1
	var top int = 0
	var bottom int = rows - 1
	var answer []int

	for left <= right && top <= bottom {
		for i := left; i <= right; i++ {
			answer = append(answer, matrix[top][i])
		}
		top++
		for i := top; i <= bottom; i++ {
			answer = append(answer, matrix[i][right])
		}
		right--
		if top <= bottom {
			for i := right; i >= left; i-- {
				answer = append(answer, matrix[bottom][i])
			}
			bottom--
		}
		if left <= right {
			for i := bottom; i >= top; i-- {
				answer = append(answer, matrix[i][left])
			}
			left++
		}
	}
	return answer
}

func ThreeSumI(nums []int, target int) [3]int {
	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			for k := j + 1; k < len(nums); k++ {
				if nums[i]+nums[j]+nums[k] == target {
					return [3]int{i, j, k}
				}
			}
		}
	}
	return [3]int{-1, -1, -1}
}

File: c/arrays.c
#include <limits.h>
#include <malloc.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

int maxInt(int a, int b) {
  if (a > b) {
    return a;
  }
  return b;
}

int minInt(int a, int b) {
  if (a < b) {
    return a;
  }
  return b;
}

void Reverse(int nums[], size_t size) {
  int i = 0;
  int j = size - 1;
  while (i < j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
    i++;
    j--;
  }
}

struct CounterItem {
    int Item;
    int Count;
};

bool SearchCounter(struct CounterItem *counter, int cSize, int key) {
  for (int i = 0; i < cSize; i++) {
    if (counter[i].Item == key) {
      return true;
    }
  }
  return false;
}

void AddCounterItem(struct CounterItem **counter, int *cSize, int key) {
  *counter = realloc(*counter, (*cSize + 1) * sizeof(struct CounterItem));
  if (*counter == NULL) {
    exit(EXIT_FAILURE);
  }
  (*counter)[*cSize].Item = key;
  (*counter)[*cSize].Count = 1;
  (*cSize)++;
}

void IncrementCounter(struct CounterItem *counter, int cSize, int key) {
  for (int i = 0; i < cSize; i++) {
    if (counter[i].Item == key) {
      counter[i].Count++;
    }
  }
}

struct CounterItem *CreateCounter(int nums[], size_t size, int *cSize) {
  struct CounterItem *counter =
      (struct CounterItem *)malloc(sizeof(struct CounterItem));
  *cSize = 0;
  for (int i = 0; i < size; i++) {
    if (SearchCounter(counter, *cSize, nums[i]) == true) {
      IncrementCounter(counter, *cSize, nums[i]);
    } else {
      AddCounterItem(&counter, cSize, nums[i]);
    }
  }
  return counter;
}

int LinearSearch(int nums[], size_t size, int key) {
  for (int i = 0; i < size; i++) {
    if (nums[i] == key) {
      return i;
    }
  }
  return -1;
}

int *SearchRange(int nums[], size_t size, int key, int *returnSize) {
  int *answer = (int *)malloc(sizeof(int) * 2);
  *returnSize = 2;
  answer[0] = -1;
  answer[1] = -1;
  for (int i = 0; i < size; i++) {
    if (nums[i] == key) {
      answer[0] = i;
      break;
    }
  }
  for (int j = size - 1; j > 0; j--) {
    if (nums[j] == key) {
      answer[1] = j;
      break;
    }
  }
  return answer;
}

int FindHighestFrequency(int nums[], size_t size) {
  int answer, count;
  count = INT_MIN;
  int cSize = 0;
  struct CounterItem *c = CreateCounter(nums, size, &cSize);
  for (int i = 0; i < cSize; i++) {
    if (c[i].Count > count) {
      answer = c[i].Item;
      count = c[i].Count;
    }
  }
  return answer;
}

int FindLowestFrequency(int nums[], size_t size) {
  int answer, count;
  count = INT_MAX;
  int cSize = 0;
  struct CounterItem *c = CreateCounter(nums, size, &cSize);
  for (int i = 0; i < cSize; i++) {
    if (c[i].Count < count) {
      answer = c[i].Item;
      count = c[i].Count;
    }
  }
  return answer;
}

int FindLargest(int nums[], size_t size) {
  int largest = nums[0];
  for (int i = 0; i < size; i++) {
    if (nums[i] > largest) {
      largest = nums[i];
    }
  }
  return largest;
}

int FindSmallest(int nums[], size_t size) {
  int smallest = nums[0];
  for (int i = 0; i < size; i++) {
    if (nums[i] < smallest) {
      smallest = nums[i];
    }
  }
  return smallest;
}

int FindSecondLargest(int nums[], size_t size) {
  int largest = nums[0];
  int secondLargest = nums[0];
  for (int i = 0; i < size; i++) {
    if (nums[i] < largest) {
      largest = nums[i];
    }
    if (nums[i] > largest && nums[i] < secondLargest) {
      secondLargest = nums[i];
    }
  }
  return secondLargest;
}

int FindSecondSmallest(int nums[], size_t size) {
  int smallest = nums[0];
  int secondSmallest = nums[0];
  for (int i = 0; i < size; i++) {
    if (nums[i] < smallest) {
      smallest = nums[i];
    }
    if (nums[i] > smallest && nums[i] < secondSmallest) {
      secondSmallest = nums[i];
    }
  }
  return secondSmallest;
}

int *Leaders(int nums[], size_t size, int *returnSize) {
  int *answer = (int *)malloc(sizeof(int) * size);
  int max = nums[size - 1];
  answer[0] = max;
  int index = 1;
  for (int i = size - 2; i >= 0; i--) {
    if (nums[i] >= max) {
      max = nums[i];
      answer[index] = nums[i];
      index++;
    }
  }
  *returnSize = index;
  Reverse(answer, *returnSize);
  return answer;
}

int FindSingleI(int nums[], size_t size) {
  int cSize = 0;
  struct CounterItem *c = CreateCounter(nums, size, &cSize);
  for (int i = 0; i < cSize; i++) {
    if (c[i].Count == 1) {
      return c[i].Item;
    }
  }
  return 0;
}

int FindSingleII(int nums[], size_t size) {
  int xorSum = 0;
  for (int i = 0; i < size; i++) {
    xorSum = xorSum ^ nums[i];
  }
  return xorSum;
}

bool CheckSorted(int nums[], size_t size) {
  for (int i = 0; i < size - 1; i++) {
    if (nums[i] > nums[i + 1]) {
      return false;
    }
  }
  return true;
}

bool CheckRotatedSorted(int nums[], size_t size) {
  int count = 0;
  for (int i = 1; i < size; i++) {
    if (nums[i] < nums[i - 1]) {
      count++;
    }
  }
  if (count == 0 || count == 1 && nums[0] > nums[size - 1]) {
    return true;
  }
  return false;
}

int FindEvenDigits(int nums[], size_t size) {
  int answer = 0;
  for (int i = 0; i < size; i++) {
    int digitCount = 0;
    int digit = nums[i];
    while (digit > 0) {
      digit = digit / 10;
      digitCount++;
    }
    if (digitCount % 2 == 0) {
      answer++;
    }
  }
  return answer;
}

int MaxConsecutiveOnes(int nums[], size_t size) {
  int localCount = 0;
  int globalCount = 0;
  for (int i = 0; i < size; i++) {
    if (nums[i] == 1) {
      localCount++;
    } else {
      localCount = 0;
    }
    globalCount = maxInt(globalCount, localCount);
  }
  return globalCount;
}

int FindMissing(int nums[], size_t size) {
  int sum = 0;
  int n = size;
  for (int i = 0; i < size; i++) {
    sum += nums[i];
  }
  return n * (n + 1) / 2 - sum;
}

int *TwoSumI(int nums[], size_t size, int target, int *returnSize) {
  int *answer = (int *)malloc(sizeof(int) * 2);
  *returnSize = 2;
  answer[0] = -1;
  answer[1] = -1;
  for (int i = 0; i < size; i++) {
    for (int j = i + 1; j < size; j++) {
      if (nums[i] + nums[j] == target) {
        answer[0] = nums[i];
        answer[1] = nums[j];
        return answer;
      }
    }
  }
  return answer;
}

int RemoveDuplicates(int nums[], size_t size) {
  int i = 0;
  for (int j = 1; j < size; j++) {
    if (nums[i] != nums[j]) {
      i++;
      nums[i] = nums[j];
    }
  }
  return i + 1;
}

int RemoveZeros(int nums[], size_t size) {
  int i = 0;
  for (int j = 0; j < size; j++) {
    if (nums[j] != 0) {
      int temp = nums[j];
      nums[j] = nums[i];
      nums[i] = temp;
      i++;
    }
  }
  return i + 1;
}

bool ContainsDuplicates(int nums[], size_t size) {
  int cSize = 0;
  struct CounterItem *c = CreateCounter(nums, size, &cSize);
  for (int i = 0; i < cSize; i++) {
    if (c[i].Count > 1) {
      return true;
    }
  }
  return false;
}

void RightRotateByOne(int nums[], size_t size) {
  if (size == 1) {
    return;
  }
  int temp = nums[size - 1];
  for (int i = size - 1; i >= 0; i--) {
    nums[i] = nums[i - 1];
  }
  nums[0] = temp;
}

void RotateByK(int nums[], size_t size, int k) {
  k = k % size;
  int i = 0;
  int j = size - 1;
  while (i <= j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
    i++;
    j--;
  }
  i = 0;
  j = k - 1;
  while (i <= j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
    i++;
    j--;
  }
  i = k;
  j = size - 1;
  while (i <= j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
    i++;
    j--;
  }
}

int LowerBound(int nums[], size_t size, int target) {
  int start = 0;
  int end = size - 1;
  int answer = size;
  while (start <= end) {
    int mid = (start + end) / 2;
    if (nums[mid] >= target) {
      answer = mid;
      end = mid - 1;
    } else {
      start = mid + 1;
    }
  }
  return answer;
}

void SortColors(int nums[], size_t size) {
  int low, mid;
  low = mid = 0;
  int high = size - 1;
  while (mid <= high) {
    if (nums[mid] == 0) {
      int temp = nums[mid];
      nums[mid] = nums[low];
      nums[low] = temp;
      low++;
      mid++;
    } else if (nums[mid] == 2) {
      int temp = nums[mid];
      nums[mid] = nums[high];
      nums[high] = temp;
      high--;
    } else {
      mid++;
    }
  }
}

int MajorityElementI(int nums[], size_t size) {
  int count = 0;
  int element = nums[0];
  for (int i = 0; i < size; i++) {
    if (nums[i] == element) {
      count++;
    } else if (count == 0) {
      element = nums[i];
      count++;
    } else {
      count--;
    }
  }
  return element;
}

int *MajorityElementII(int nums[], size_t size, int *returnSize) {
  int c1 = 0;
  int c2 = 0;
  int e1 = 0;
  int e2 = 0;
  for (int i = 0; i < size; i++) {
    if (nums[i] == e1) {
      c1++;
    } else if (nums[i] == e2) {
      c2++;
    } else if (c1 == 0) {
      e1 = nums[i];
      c1 = 1;
    } else if (c2 == 0) {
      e2 = nums[i];
      c2 = 1;
    } else {
      c1--;
      c2--;
    }
  }
  c1 = 0;
  c2 = 0;
  for (int i = 0; i < size; i++) {
    if (nums[i] == e1) {
      c1++;
    } else if (nums[i] == e2) {
      c2++;
    }
  }
  int *answer = (int *)malloc(sizeof(int) * 2);
  *returnSize = 0;
  if (c1 > size / 3) {
    answer[(*returnSize)++] = e1;
  }
  if (c2 > size / 3) {
    answer[(*returnSize)++] = e2;
  }
  return answer;
}

int MaxSubarraySumI(int nums[], size_t size) {
  int globalSum = INT_MIN;
  int localSum = 0;
  for (int i = 0; i < size; i++) {
    localSum = maxInt(nums[i], nums[i] + localSum);
    globalSum = maxInt(globalSum, localSum);
  }
  return globalSum;
}

int *MaxSubarraySumII(int nums[], size_t size, int *returnSize) {
  int globalSum = INT_MIN;
  int localSum = 0;
  int start = 0;
  int end = 0;
  *returnSize = 2;
  for (int i = 0; i < size; i++) {
    if (localSum == 0) {
      start = i;
    }
    localSum += nums[i];
    if (localSum > globalSum) {
      globalSum = localSum;
      end = i;
    }
    if (localSum <= 0) {
      localSum = 0;
    }
  }
  int *answer = (int *)malloc(sizeof(int) * 2);
  answer[0] = start;
  answer[1] = end;
  return answer;
}

int MaxProfit(int *prices, int pricesSize) {
  int diff = -1;
  int min = INT_MAX;
  for (int i = 0; i < pricesSize; i++) {
    min = minInt(min, prices[i]);
    diff = maxInt(diff, prices[i] - min);
  }
  return diff;
}

int *RearrangeBySign(int *nums, int size, int *returnSize) {
  int *answer = (int *)malloc(sizeof(int) * size);
  *returnSize = size;
  int positive = 0;
  int negative = 1;
  for (int i = 0; i < size; i++) {
    if (nums[i] >= 0) {
      answer[positive] = nums[i];
      positive += 2;
    } else {
      answer[negative] = nums[i];
      negative += 2;
    }
  }
  return answer;
}

void Merge(int nums1[], int nums2[], size_t size1, size_t size2) {
  int i = size1 - 1;
  int j = size2 - 1;
  int k = size1 + size2 - 1;
  while (i >= 0 && j >= 0) {
    if (nums1[i] > nums2[j]) {
      nums1[k] = nums1[i];
      i--;
    } else {
      nums1[k] = nums2[j];
      j--;
    }
    k--;
  }
  while (j >= 0) {
    nums1[k] = nums2[j];
    j--;
    k--;
  }
}
